SwiftVoiceQuickstart/ViewController.swift

```diff
@@ -453,6 +454,82 @@ extension ViewController: TVOCallDelegate {
+        /**
+        * currentWarnings: existing quality warnings that have not been cleared yet
+        * previousWarnings: last set of warnings prior to receiving this callback
+        *
+        * Example:
+        *   - currentWarnings: { A, B }
+        *   - previousWarnings: { B, C }
+        *   - intersection: { B }
+        *
+        * Newly raised warnings = currentWarnings - intersection = { A }
+        * Newly cleared warnings = previousWarnings - intersection = { C }
+        */
+        var warningsIntersection: Set<NSNumber> = currentWarnings
+        warningsIntersection = warningsIntersection.intersection(previousWarnings)
+        
+        var newWarnings: Set<NSNumber> = currentWarnings
+        newWarnings.subtract(warningsIntersection)
+        if newWarnings.count > 0 {
+            qualityWarningsUpdatePopup(newWarnings, isCleared: false)
+        }
+        
+        var clearedWarnings: Set<NSNumber> = previousWarnings
+        clearedWarnings.subtract(warningsIntersection)
+        if clearedWarnings.count > 0 {
+            qualityWarningsUpdatePopup(clearedWarnings, isCleared: true)
+        }
+    }
+    
+    func qualityWarningsUpdatePopup(_ warnings: Set<NSNumber>, isCleared: Bool) {
+        var popupMessage: String! = "Warnings detected:"
+        if isCleared {
+            popupMessage = "Warnings cleared:"
+        }
+        
+        for warning in warnings {
+            if let warningName = warningString(TVOCallQualityWarning(rawValue: warning.uintValue)!) {
+                popupMessage = popupMessage + " " + warningName
+            }
+        }
+        
+        qualityWarningsToaster.alpha = 0.0
+        qualityWarningsToaster.text = popupMessage
+        UIView.animate(withDuration: 1.0, animations: {
+            self.qualityWarningsToaster.isHidden = false
+            self.qualityWarningsToaster.alpha = 1.0
+        }) { [weak self] finish in
+            guard let strongSelf = self else { return }
+            let deadlineTime = DispatchTime.now() + .seconds(5)
+            DispatchQueue.main.asyncAfter(deadline: deadlineTime, execute: {
+                UIView.animate(withDuration: 1.0, animations: {
+                    strongSelf.qualityWarningsToaster.alpha = 0.0
+                }) { (finished) in
+                    strongSelf.qualityWarningsToaster.isHidden = true
+                }
+            })
+        }
+    }
+    
+    func warningString(_ warning: TVOCallQualityWarning) -> String! {
+        switch warning {
+        case .highRtt:
+            return "high-rtt"
+        case .highJitter:
+            return "high-jitter"
+        case .highPacketsLostFraction:
+            return "high-packets-lost-fraction"
+        case .lowMos:
+            return "low-mos"
+        case .constantAudioInputLevel:
+            return "constant-audio-input-level"
+        default:
+            return "Unknown warning"
+        }
+    }
```

Comment:
You could do something like this to move code out of this large class and be more single responsibility. Makes usage pretty nice also.
```swift
private extension TVOCallQualityWarning {
    var message: String {
        switch self {
            case .highRtt: return "high-rtt"
            // ...
        }
    }
}
```
Single line cases can make simple switches like this easier to read also.

Reply:
Also, all of the work being done to compose the complete message (not just this switch) is really just data manipulation which is a great opportunity to move code out of the large controller file because it really is separate from UI work. This can be done with a factory or view model pattern. This PR probably is not the time but maybe it's something I could talk about in iOS guild meeting because there are some tricks.